File: day2/day2.py

def parse_input(input):
    direction, magnitude = input.split(" ")
    return direction, int(magnitude)

unitary_map = {
    'forward': (1, 0),
    'backward': (-1, 0),
    'up': (0, -1),
    'down': (0, 1)
}

def project(direction_vector, magnitude):
    """Does not normalise, only works on unitary vecs"""
    output = list() 
    for dim in direction_vector:
        output.append(dim*magnitude)
    return output

def move(starting_coords, movement):
    output = list()
    for s, m in zip(starting_coords, movement):
        output.append(s + m)
    return output

def vector_product(vector):
    running_prod = 1
    for v in vector:
        running_prod = running_prod * v
    return running_prod


def question1(inputs):
    
    current_coords = [0,0]

    for instruction in inputs:
        direction, magnitude = parse_input(instruction)
        movement = project(unitary_map[direction], magnitude)
        current_coords = move(current_coords, movement)
    return vector_product(current_coords)


def move_q2(starting_coords, movement, magnitude):
    output = list()

    # Forwards
    output.append(starting_coords[0] + movement[0]*magnitude)

    # Up/ down
    output.append(starting_coords[1] + starting_coords[2]*movement[0]*magnitude)

    # aim
    output.append(starting_coords[2] + movement[1]*magnitude)

    return output


def question2(inputs):

    #coords: x,y,aim
    current_coords = [0,0,0]
    for instruction in inputs:
        direction, magnitude = parse_input(instruction)
        current_coords = move_q2(current_coords, unitary_map[direction], magnitude)
    return vector_product(current_coords[:2])


if __name__ == "__main__":
    with open("day2/inputs/test1.txt", "r") as f:
        test1 = f.readlines()
        test1 = [t.replace("\n", "") for t in test1]

    with open("day2/inputs/input1.txt", "r") as f:
        input1 = f.readlines()
        input1 = [t.replace("\n", "") for t in input1]

    print(question1(input1))
    print(question2(input1))

File: day2/day2.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

func Reader(r io.Reader) ([]string, error) {
	scanner := bufio.NewScanner(r)
	scanner.Split(bufio.ScanLines)
	var result []string
	for scanner.Scan() {
		x := scanner.Text()
		result = append(result, x)
	}
	return result, scanner.Err()
}

// Part 1
func MoveSubmarine(starting_position [2]int, instruction string) [2]int {
	instruction_vector := strings.Fields(instruction)
	direction := instruction_vector[0]
	units, _ := strconv.Atoi(instruction_vector[1])

	var output [2]int
	if direction == "forward" {
		output[0], output[1] = starting_position[0]+units, starting_position[1]
	} else if direction == "up" {
		output[0], output[1] = starting_position[0], starting_position[1]-units
	} else if direction == "down" {
		output[0], output[1] = starting_position[0], starting_position[1]+units
	}

	return output
}

// Part 2
func MoveSubmarinePart2(starting_position [3]int, instruction string) [3]int {
	instruction_vector := strings.Fields(instruction)
	direction := instruction_vector[0]
	units, _ := strconv.Atoi(instruction_vector[1])

	output := starting_position

	if direction == "forward" {
		output[0] = starting_position[0] + units
		output[1] = starting_position[1] + starting_position[2]*units
	} else if direction == "up" {
		output[2] = starting_position[2] - units
	} else if direction == "down" {
		output[2] = starting_position[2] + units
	}

	return output
}

func main() {
	file, _ := os.Open("day2/inputs/input1.txt")
	input, _ := Reader(file)

	// Part 1
	position := [2]int{0, 0} //start at 0,0

	for _, instruction := range input {
		position = MoveSubmarine(position, instruction)
	}

	fmt.Println(position)
	fmt.Println(position[0] * position[1])

	//Part 2
	position2 := [3]int{0, 0, 0} //start at 0,0,0, (x,y,aim)

	for _, instruction := range input {
		position2 = MoveSubmarinePart2(position2, instruction)
	}

	fmt.Println(position2)
	fmt.Println(position2[0] * position2[1])

}
